sol/Usdx.sol

```diff
@@ -1,48 +1,148 @@
- *   - Received eth is then burned forever.
+ *   - Eth sent to the usdx contract is locked, and USDX is minted
+ *     into the sender's account at the current eth/usd exchange rate.
+ *   - Locked eth can be redeemed by the original sender by burning USDX
+ *     at the originally minted price.
  *
- *   Note: Owner is able to set the Eth/USD oracle.
+ *   Note: Owner is able to set the eth/usd oracle.
  */
 contract USDX is ERC20, Ownable {
 	using SafeMath for uint256;
 	using SafeCast for int256;
 	uint8 constant FEED_DECS = 8;
-	// ETH/USD exchange rate feed.
-	AggregatorV3Interface public ethUsdPriceFeed;
+	AggregatorV3Interface public usdPriceFeed;
+	uint8 public feedDecs;
 
-	constructor (address _priceFeed) public ERC20("USDX Stablecoin", "USDX") {
+	struct account {
+		uint256 locked;
+		uint256 mint;
+	}
+	mapping (address => account) public accounts;
+
+	constructor (address _priceFeed) ERC20("USDX Stablecoin", "USDX") {
 		setFeed(_priceFeed);
 	}
 
+	// _newFeed's decimals parameter must immutably be set to 8.
 	function setFeed(address _newFeed) public onlyOwner {
 		AggregatorV3Interface newFeed = AggregatorV3Interface(_newFeed);
 		require(newFeed.decimals() == FEED_DECS);
-		ethUsdPriceFeed = newFeed;
+		usdPriceFeed = newFeed;
 	}
 
+	// Received eth is minted into usdx at the current eth/usd
+	// exchange rate.
+	// Note: msg.sender must be payable to allow redemption of
+	// deposited eth.
 	receive() external payable {
-		(,int256 rate,,,) = ethUsdPriceFeed.latestRoundData();
-		uint256 usd = msg.value.mul(rate.toUint256()).div(10**FEED_DECS); //latestRoundData uses 8 decimals
-		_mint(_msgSender(), usd);
+		(,int256 rate,,,) = usdPriceFeed.latestRoundData();
+		uint256 toMint = weiToUSDX(msg.value, rate);
+		account storage acct = accounts[msg.sender];
+		acct.locked += msg.value;
+		acct.mint += toMint;
+		_mint(msg.sender, toMint);
+	}
+
+	// Redeems usdx back into eth.  This method can only be called by
+	// senders who have previously directly sent eth into this
+	// contract.  The amount of eth redeemed is capped to the amount
+	// of eth the sender has previously sent.  The amount of eth
+	// redeemed is unrelated to the current ETH/USD price; it's based
+	// purely on the ratio of previously sent eth and previously
+	// minted usdx (ie: if 1 eth was previously received by this
+	// contract to mint 1000usdx, 500 usdx is able to be redeemed for
+	// .5 eth regardless of whether usd/eth price has decreased to
+	// 800usd/eth).  If _amount is 0, msg.sender's full usdx balance
+	// will be used to redeem eth.  Otherwise, _amount is the maximum
+	// amount of usdx used to redeem into eth.
+	function redeem(uint256 _amount) public returns (uint256) {
+		account storage acct = accounts[msg.sender];
+		require(acct.locked > 0, "nothing to redeem");
+		if (_amount == 0) {
+			_amount = acct.mint;
+		} else {
+			_amount = min(_amount, acct.mint);
+		}
+
+		_amount = min(_amount, balanceOf(msg.sender));
```

Comment:
Can balanceOf(msg.sender) be different than acct.mint, or should these be equivalent?

Reply:
Yes, they can be different.  `acct.mint` is the amount of USDX minted by sender, while `balanceOf` is current USDX balance which can fluctuate due to normal ERC20 transfers (either sending USDX to other accounts, or receiving USDX from others).

This code protects against the following case:
- Mint 1000usdx for 1 eth
- Send 600usdx to other account (current balance 400)
- Redeem.
```
acct.mint == 1000
balanceOf == 400
```
In this case, user should only be allowed to redeem 400 usdx worth of eth.

Reply:
I see, makes sense now. Thanks for explaining